File: matrix.py
def input_matrix():
    rows = int(input("Masukkan jumlah baris: "))
    cols = int(input("Masukkan jumlah kolom: "))

    matrix = []
    print("Masukkan elemen matriks:")

    for i in range(rows):
        row = list(map(int, input(f"Baris {i+1}: ").split()))
        matrix.append(row)

    return matrix


def print_matrix(matrix):
    print("\nMatriks:")
    for row in matrix:
        print(*row)
    print()


# 1. Sort Row Wise
def sort_row_wise(matrix):
    return [sorted(row) for row in matrix]


# 2. Sort Column Wise
def sort_column_wise(matrix):
    rows = len(matrix)
    cols = len(matrix[0])

    for j in range(cols):
        col = [matrix[i][j] for i in range(rows)]
        col.sort()

        for i in range(rows):
            matrix[i][j] = col[i]

    return matrix


# 3. Rotate Clockwise by 1
def rotate_clockwise_by_one(matrix):
    rows = len(matrix)
    cols = len(matrix[0])

    temp = matrix[0][0]

    for i in range(rows - 1):
        matrix[i][0] = matrix[i + 1][0]

    for j in range(cols - 1):
        matrix[rows - 1][j] = matrix[rows - 1][j + 1]

    for i in range(rows - 1, 0, -1):
        matrix[i][cols - 1] = matrix[i - 1][cols - 1]

    for j in range(cols - 1, 1, -1):
        matrix[0][j] = matrix[0][j - 1]

    matrix[0][1] = temp
    return matrix


# 4. Rotate Counter Clockwise by 1
def rotate_counter_clockwise_by_one(matrix):
    rows = len(matrix)
    cols = len(matrix[0])

    temp = matrix[0][0]

    for j in range(cols - 1):
        matrix[0][j] = matrix[0][j + 1]

    for i in range(rows - 1):
        matrix[i][cols - 1] = matrix[i + 1][cols - 1]

    for j in range(cols - 1, 0, -1):
        matrix[rows - 1][j] = matrix[rows - 1][j - 1]

    for i in range(rows - 1, 1, -1):
        matrix[i][0] = matrix[i - 1][0]

    matrix[1][0] = temp
    return matrix


# 5. Rotate 90 Derajat
def rotate_90(matrix):
    return [list(row) for row in zip(*matrix[::-1])]


# 6. Rotate 180 Derajat
def rotate_180(matrix):
    return [row[::-1] for row in matrix[::-1]]


# 7. Row Wise Traversal
def row_wise_traversal(matrix):
    print("Row-wise Traversal:")
    for row in matrix:
        for item in row:
            print(item, end=" ")
    print("\n")


# 8. Column Wise Traversal
def column_wise_traversal(matrix):
    print("Column-wise Traversal:")

    rows = len(matrix)
    cols = len(matrix[0])

    for j in range(cols):
        for i in range(rows):
            print(matrix[i][j], end=" ")

    print("\n")


# 9. Spiral Print
def spiral_print(matrix):
    print("Spiral Traversal:")

    top = 0
    bottom = len(matrix) - 1
    left = 0
    right = len(matrix[0]) - 1

    while top <= bottom and left <= right:

        for i in range(left, right + 1):
            print(matrix[top][i], end=" ")
        top += 1

        for i in range(top, bottom + 1):
            print(matrix[i][right], end=" ")
        right -= 1

        if top <= bottom:
            for i in range(right, left - 1, -1):
                print(matrix[bottom][i], end=" ")
            bottom -= 1

        if left <= right:
            for i in range(bottom, top - 1, -1):
                print(matrix[i][left], end=" ")
            left += 1

    print("\n")


# 10. Transpose Matrix
def transpose_matrix(matrix):
    return [list(row) for row in zip(*matrix)]


# =====================
# MAIN PROGRAM
# =====================

matrix_asli = input_matrix()

while True:
    print("===== MENU MATRIX =====")
    print("1. Sort Matrix Row-wise")
    print("2. Sort Matrix Column-wise")
    print("3. Rotate Matrix Clockwise by 1")
    print("4. Rotate Matrix Counter-Clockwise by 1")
    print("5. Rotate Matrix 90°")
    print("6. Rotate Matrix 180°")
    print("7. Row-wise Traversal")
    print("8. Column-wise Traversal")
    print("9. Print Matrix in Spiral Form")
    print("10. Transpose Matrix")
    print("11. Quit")

    choice = input("Pilih menu: ")

    if choice == "1":
        hasil = sort_row_wise([row[:] for row in matrix_asli])
        print_matrix(hasil)

    elif choice == "2":
        hasil = sort_column_wise([row[:] for row in matrix_asli])
        print_matrix(hasil)

    elif choice == "3":
        hasil = rotate_clockwise_by_one([row[:] for row in matrix_asli])
        print_matrix(hasil)

    elif choice == "4":
        hasil = rotate_counter_clockwise_by_one([row[:] for row in matrix_asli])
        print_matrix(hasil)

    elif choice == "5":
        hasil = rotate_90([row[:] for row in matrix_asli])
        print_matrix(hasil)

    elif choice == "6":
        hasil = rotate_180([row[:] for row in matrix_asli])
        print_matrix(hasil)

    elif choice == "7":
        row_wise_traversal(matrix_asli)

    elif choice == "8":
        column_wise_traversal(matrix_asli)

    elif choice == "9":
        spiral_print(matrix_asli)

    elif choice == "10":
        hasil = transpose_matrix([row[:] for row in matrix_asli])
        print_matrix(hasil)

    elif choice == "11":
        print("Program selesai.")
        break

    else:
        print("Pilihan tidak valid!")

File: Matrix.java
import java.util.Scanner;
import java.util.Arrays;

public class Matrix {

    static Scanner input = new Scanner(System.in);

    public static void printMatrix(int[][] matrix) {
        System.out.println("\nMatrix:");
        for (int[] row : matrix) {
            for (int value : row) {
                System.out.print(value + "\t");
            }
            System.out.println();
        }
    }

    // 1-a Sort Row Wise
    public static void sortRowWise(int[][] matrix) {
        for (int i = 0; i < matrix.length; i++) {
            Arrays.sort(matrix[i]);
        }
    }

    // 1-b Sort Column Wise
    public static void sortColumnWise(int[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;

        for (int j = 0; j < cols; j++) {
            int[] temp = new int[rows];

            for (int i = 0; i < rows; i++) {
                temp[i] = matrix[i][j];
            }

            Arrays.sort(temp);

            for (int i = 0; i < rows; i++) {
                matrix[i][j] = temp[i];
            }
        }
    }

    // 2-a Rotate Clockwise by 1
    public static void rotateClockwiseByOne(int[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;

        if (rows < 2 || cols < 2)
            return;

        int temp = matrix[0][0];

        for (int i = 0; i < rows - 1; i++)
            matrix[i][0] = matrix[i + 1][0];

        for (int j = 0; j < cols - 1; j++)
            matrix[rows - 1][j] = matrix[rows - 1][j + 1];

        for (int i = rows - 1; i > 0; i--)
            matrix[i][cols - 1] = matrix[i - 1][cols - 1];

        for (int j = cols - 1; j > 1; j--)
            matrix[0][j] = matrix[0][j - 1];

        matrix[0][1] = temp;
    }

    // 2-b Rotate Counter Clockwise by 1
    public static void rotateCounterClockwiseByOne(int[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;

        if (rows < 2 || cols < 2)
            return;

        int temp = matrix[0][0];

        for (int j = 0; j < cols - 1; j++)
            matrix[0][j] = matrix[0][j + 1];

        for (int i = 0; i < rows - 1; i++)
            matrix[i][cols - 1] = matrix[i + 1][cols - 1];

        for (int j = cols - 1; j > 0; j--)
            matrix[rows - 1][j] = matrix[rows - 1][j - 1];

        for (int i = rows - 1; i > 1; i--)
            matrix[i][0] = matrix[i - 1][0];

        matrix[1][0] = temp;
    }

    // 2-c Rotate 90°
    public static int[][] rotate90(int[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;

        int[][] result = new int[cols][rows];

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                result[j][rows - 1 - i] = matrix[i][j];
            }
        }

        return result;
    }

    // 2-d Rotate 180°
    public static int[][] rotate180(int[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;

        int[][] result = new int[rows][cols];

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                result[rows - 1 - i][cols - 1 - j] = matrix[i][j];
            }
        }

        return result;
    }

    // 3-a Row Wise Traversal
    public static void rowWiseTraversal(int[][] matrix) {
        System.out.println("Row Wise Traversal:");
        for (int[] row : matrix) {
            for (int value : row) {
                System.out.print(value + " ");
            }
        }
        System.out.println();
    }

    // 3-b Column Wise Traversal
    public static void columnWiseTraversal(int[][] matrix) {
        System.out.println("Column Wise Traversal:");

        for (int j = 0; j < matrix[0].length; j++) {
            for (int i = 0; i < matrix.length; i++) {
                System.out.print(matrix[i][j] + " ");
            }
        }
        System.out.println();
    }

    // 4 Spiral Form
    public static void spiralPrint(int[][] matrix) {
        int top = 0;
        int bottom = matrix.length - 1;
        int left = 0;
        int right = matrix[0].length - 1;

        System.out.println("Spiral Traversal:");

        while (top <= bottom && left <= right) {

            for (int i = left; i <= right; i++)
                System.out.print(matrix[top][i] + " ");
            top++;

            for (int i = top; i <= bottom; i++)
                System.out.print(matrix[i][right] + " ");
            right--;

            if (top <= bottom) {
                for (int i = right; i >= left; i--)
                    System.out.print(matrix[bottom][i] + " ");
                bottom--;
            }

            if (left <= right) {
                for (int i = bottom; i >= top; i--)
                    System.out.print(matrix[i][left] + " ");
                left++;
            }
        }

        System.out.println();
    }

    // 5 Transpose
    public static int[][] transpose(int[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;

        int[][] result = new int[cols][rows];

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                result[j][i] = matrix[i][j];
            }
        }

        return result;
    }

    public static void main(String[] args) {

        System.out.print("Jumlah baris: ");
        int rows = input.nextInt();

        System.out.print("Jumlah kolom: ");
        int cols = input.nextInt();

        int[][] matrix = new int[rows][cols];

        System.out.println("Masukkan elemen matrix:");

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                matrix[i][j] = input.nextInt();
            }
        }

        int pilihan;

        do {
            System.out.println("\n===== MENU MATRIX =====");
            System.out.println("1. Sort Matrix Row-wise");
            System.out.println("2. Sort Matrix Column-wise");
            System.out.println("3. Rotate Matrix Clockwise by 1");
            System.out.println("4. Rotate Matrix Counter-Clockwise by 1");
            System.out.println("5. Rotate Matrix 90°");
            System.out.println("6. Rotate Matrix 180°");
            System.out.println("7. Row-wise Traversal");
            System.out.println("8. Column-wise Traversal");
            System.out.println("9. Print Matrix in Spiral Form");
            System.out.println("10. Transpose Matrix");
            System.out.println("11. Quit");
            System.out.print("Pilih menu: ");

            pilihan = input.nextInt();

            switch (pilihan) {

                case 1:
                    sortRowWise(matrix);
                    printMatrix(matrix);
                    break;

                case 2:
                    sortColumnWise(matrix);
                    printMatrix(matrix);
                    break;

                case 3:
                    rotateClockwiseByOne(matrix);
                    printMatrix(matrix);
                    break;

                case 4:
                    rotateCounterClockwiseByOne(matrix);
                    printMatrix(matrix);
                    break;

                case 5:
                    matrix = rotate90(matrix);
                    printMatrix(matrix);
                    break;

                case 6:
                    matrix = rotate180(matrix);
                    printMatrix(matrix);
                    break;

                case 7:
                    rowWiseTraversal(matrix);
                    break;

                case 8:
                    columnWiseTraversal(matrix);
                    break;

                case 9:
                    spiralPrint(matrix);
                    break;

                case 10:
                    matrix = transpose(matrix);
                    printMatrix(matrix);
                    break;

                case 11:
                    System.out.println("Program selesai.");
                    break;

                default:
                    System.out.println("Pilihan tidak valid!");
            }

        } while (pilihan != 11);
    }
}